### ISSUE TITLE
Alt/super+tab recency-based cycling

### ISSUE THREAD

**Issue by WatchMkr** [pop-os/cosmic-launcher#135]
**Recency-Based Cycling:** The Alt+Tab interface should prioritize switching between applications based on the order of their last use. This means that the most recently used application will always be the first in the cycle, followed by the second most recently used, and so on.

**1. Primary Use Case:** When Alt+Tab is pressed and released quickly, it should switch directly to the last used application.
**2. Extended Use Case:** Holding Alt and tapping Tab repeatedly should cycle through the list of open applications in the order of their recency. This allows users to quickly access any of the last few applications they were working with.

**Comment by darklajid** [comment]
@WatchMkr would it make sense to support shift-alt-tab (cycle backwards) too in the same requirement?
Would this be something that outsiders could contribute to?

**Comment by WatchMkr** [comment]
@darklajid Yes, shift-alt-tab should cycle backwards. We'd welcome a PR.

**Comment by curiousercreative** [comment]
@WatchMkr is there a story for increasing the number of windows that can be cycled through? I always have more than 8 windows open in my work day and it's been frustrating to have to rely on mouse and dock to find a window when it's not in the list.

**Comment by thedeanominator** [comment]
Regarding the use of ALT + TAB on PopOS (regardless of whether using the current alpha version of COSMIC) to switch between windows, is there a particular logic behind the order of the applications listed? It doesn't appear to be in alphabetical order.

**Comment by WatchMkr** [comment]
@curiousercreative the plan is to remove the 8 entry limitation when in tab+alt/super mode.

**Comment by Pandapip1** [comment]
If this is implemented, there should be an option to hide the focused window from the selector.

**Comment by PoisonFrog** [comment]
> If this is implemented, there should be an option to hide the focused window from the selector.

Why? I have never seen any other desktop environment do something like this. Usually the focused window is the last one in the list. That's how Windows and every other Linux DE I know works. Why would anyone want to hide the focused windows? Sounds like an unheard-of feature that only you would use.

**Comment by Pandapip1** [comment]
> Why? I have never seen any other desktop environment do something like this. Usually the focused window is the last one in the list. That's how Windows and every other Linux DE I know works. Why would anyone want to hide the focused windows? Sounds like an unheard-of feature that only you would use.

My bad, I didn't realize that the focused window was the last one in the list. I'm just used to alt-tabbing once to switch windows, and as currently stated, tab would need to be pressed twice. Having the focused window at the end is fine for my use-case too.

**Comment by PoisonFrog** [comment]
> > Usually the focused window is the last one in the list.
> 
> Really? On GNOME I see the focused window as the first, but the initial selection would be on the second window.

Yeah, that's what I meant. The last one to be selected, unlike how it works in Cosmic where it actually is the first to be selected.

**Comment by sounddrill31** [comment]
Hope this pans out well, thanks a lot to the team for considering to fix it.

I really love cosmic so far, and it has earned its spot on my laptop - despite being in alpha. This is one of the issues that is generally workflow-breaking for me(along with a lack of a shortcut that pulls up the clipboard manager but I see that there is already an issue in the respective repo) 

Thanks a lot for this project overall

**Comment by Cedric-Lefebvre** [comment]
Hi, any news on this issue? It makes it tough to use cosmic daily with the current alt tab behaviour.

**Comment by wiiznokes** [comment]
> Hi, any news on this issue? It makes it tough to use cosmic daily with the current alt tab behaviour.

I agree. I made an initial implementation but i don't have time to work on it anymore. Here are the PRs if someone want to work on this
- https://github.com/pop-os/cosmic-launcher/pull/150
- https://github.com/pop-os/launcher/pull/227

And a related PR in cosmic comp
- https://github.com/pop-os/cosmic-comp/pull/889

**Comment by pwall2222** [comment]
#225 Should close this issue!

**Comment by Tommimon** [comment]
This implementation doesn't work properly with stacked windows. Windows in the same stack are always consecutive in the list and not recency-based ordered but stay in a predefined order.

I don't know if this is intended but it is confusing and should be changed, let me explain why:
If you have windows A and B in the same stack, let's say A is always immediately before B in the list, from A pressing Alt+Tab will bring you to B but from B pressing Alt+Tab will keep you on the same window.

This is inconsistent and not practical, Alt+Tab should never keep you on the same window unless it is the only one.

Furthermore, I don't think grouping windows in the same stack is a good idea, it brakes the behavior of this sequence of actions:
1) Alt+Tab
2) Release everything
3) Alt+Tab
We are used to assume this sequence will bring you back where you started, but currently it is not the case, this is a convention in almost every other environment and also very useful when your workflow alternates between two windows.

In conclusion I suggest to avoid grouping stacked windows at all and order all the windows by recency ignoring stacks, I believe this is how it works currently in Pop!_OS. If you want to move inside the same stack there are the "Focus window left" and "Focus window right" shortcuts already.

**Comment by snaggen** [comment]
Just installed the fedora cosmic packages version alpha.5.1-1 and cannot see any recency based behaviour for Alt-Tab.

**Comment by hgaiser** [comment]
Seems to work in Arch, also alpha 5.

**Comment by snaggen** [comment]
Seems to be an old pop-launcher in the fedora packaging... took that from the nightly copr and it seems to work.

**Comment by eevmanu** [comment]
Hi @wash2 👋🏼,

I noticed your comments in https://github.com/pop-os/cosmic-launcher/issues/130 regarding the limitation of alt-tab functionality to windows within the current workspace. You brought this up twice - once in https://github.com/pop-os/cosmic-launcher/issues/130#issuecomment-2129791864 and again in https://github.com/pop-os/cosmic-launcher/issues/130#issuecomment-2130173377 .

Could you kindly confirm whether this feature has been implemented and included in the latest updates? If not, is there a roadmap or a specific issue where this is being tracked?

Thanks a ton for your incredible work on this - it’s truly appreciated! 🎉

**Comment by mirfilip** [comment]
I'm on `cosmic-launcher/noble,now 0.1.0~1744762023~24.04~77ca208 amd64 [installed,automatic]` and it feels like this update introduced some regression or at very least changed the behaviour of the launcher significantly:

1. When focused on window A, pressing ALT+TAB and choosing window B causes window A to become a last entry of the list (e.g. has a shortcut of CTRL+9)
2. Pressing ALT+TAB continously doesn't toggle between last two used windows (as it used to) but rather toggles through the whole list. 
3. When pressing ALT+TAB and keeping ALT pressed, the focus is always on the 2nd item on the list (the one with CTRL + 2). When ALT is released, expectedly it's the one chosen. This causes what I described in point 2.

As an extra question, are all these behavious configurable in Cosmic?

**Comment by hgaiser** [comment]
> I'm on `cosmic-launcher/noble,now 0.1.0~1744762023~24.04~77ca208 amd64 [installed,automatic]` and it feels like this update introduced some regression or at very least changed the behaviour of the launcher compared to previous version. Also, it doesn't fit the description of this issue.
> 
>     1. When focused on window A, pressing ALT+TAB and choosing window B causes window A to become a last entry of the list (e.g. has a shortcut of CTRL+9)
> 
>     2. Pressing ALT+TAB continously doesn't toggle between last two used windows (as it used to) but rather toggles through the whole list.
> 
>     3. When pressing ALT+TAB and keeping ALT pressed, the focus is always on the 2nd item on the list (the one with CTRL + 2). When ALT is released, expectedly then 2nd item is chosen. This causes what I described in point 2.
> 
> 
> As an extra question, are all these behavious configurable in Cosmic?

I don't see this behavior on alpha 7. For me, repeatedly pressing alt+tab correctly switches between the last two applications.

There is one bug worth mentioning, sometimes if I release alt+tab, the application doesn't switch and the application switcher remains active (even when no keys are pressed). This seems to happen only when pressing and releasing alt+tab quickly.

**Comment by mirfilip** [comment]
> I don't see this behavior on alpha 7. For me, repeatedly pressing alt+tab correctly switches between the last two applications.

Just to clear up. When I started with pure Alpha 7, it was indeed correct. Coherent with what is described in the issue description. However, when I later did rolling updates, it has changed. 
`cosmic-launcher/noble,now 0.1.0~1744762023~24.04~77ca208` seem to have introduced it.

**Comment by moore-bryan** [comment]
This is not working correctly on 1.0.0~alpha.7^git20250623.84b1652, as alt+tab only acts as if hitting tab.
